### ISSUE TITLE
DecompilerService

### ISSUE THREAD

**Issue by astrelsky** [NationalSecurityAgency/ghidra#2017]
**Is your feature request related to a problem? Please describe.**
There is no simple way to get at the decompiler classes in use by the DecompilePlugin. At the moment if one wishes to decompile a function, outside of the FlatDecompilerAPI, you need to set up a new DecompInterface, open the program, get the results, and then release it when you're done. This starts a new decompiler process every time when there is already one currently running. 

**Describe the solution you'd like**
It would be useful to have a service from which things such as an active DecompInterface, decompiler caches, etc without having to start and release a new process and re-decompile the same function every time.

**Describe alternatives you've considered**
Stealing all fields I need via reflection

**Additional context**
After jumping through a ton of hoops I managed to grab the decompiler cache from the DecompilerController used in the DecompilerProvider via reflection. I then used it through the asMap method (provides a thread-safe view) to cache results. While I don't have much statistics, after analyzing the parameters of a functions called functions from its ClangStatements, of several hundreds of functions, the analysis time went from ~30 minutes without the cache to ~7.2 minutes with the cache.

Related to #1907

**Comment by dragonmacher** [comment]
Just to clarify, are you suggesting we provide a **pool** of Decompilers, accessible from a serivice, so that clients can decompile while avoiding the overhead of the repeated native process startup and shutdown?

For those who may not be aware, we do have tools for performing bulk decompile operations.   See the `ParallelDecompiler`.  Example uses of this class can be found in many analyzers and scripts, such as the `CompareFunctionSizesScript` and the `DecompilerSwitchAnalyzer`.

Within the `ParallelDecompiler` you can call `createChunkingParallelDecompiler()`.  This will return an interface that you can repeatedly call with functions to decompile as desired.   **Note: you must call dispose on the returned `ChunkingParallelDecompiler` when you are done.**  The other methods in the `ParallelDecompiler` will perform dispose for you, as they assume you are issuing a one-time request to decompile a group of functions (such as is done by our analyzers).

While this is not exactly a pre-existing pool of decompilers, it does allow you to achieve a similar speed-up in performance.

**Comment by astrelsky** [comment]
> Just to clarify, are you suggesting we provide a **pool** of Decompilers, accessible from a serivice, so that clients can decompile while avoiding the overhead of the repeated native process startup and shutdown?
> 
> For those who may not be aware, we do have tools for performing bulk decompile operations. See the `ParallelDecompiler`. Example uses of this class can be found in many analyzers and scripts, such as the `CompareFunctionSizesScript` and the `DecompilerSwitchAnalyzer`.
> 
> Within the `ParallelDecompiler` you can call `createChunkingParallelDecompiler()`. This will return an interface that you can repeatedly call with functions to decompile as desired. **Note: you must call dispose on the returned `ChunkingParallelDecompiler` when you are done.** The other methods in the `ParallelDecompiler` will perform dispose for you, as they assume you are issuing a one-time request to decompile a group of functions (such as is done by our analyzers).
> 
> While this is not exactly a pre-existing pool of decompilers, it does allow you to achieve a similar speed-up in performance.

Not exactly but it is a start. Ideally they should be sharing a cache similar to what is used in `DecompilerController`. This would greatly improve performance especially for larger functions which take time to decompile.

I've been doing some testing with the following. It is messy as it is still experimental and I'm still working out how to achieve what I need. However, the use of both the cache and sharing the `DecompilerInterface` has reduced analysis time from ~30 minutes to ~2 minutes.

https://github.com/astrelsky/Ghidra-Cpp-Class-Analyzer/blob/experimental/src/main/java/cppclassanalyzer/decompiler/DecompilerAPI.java

**Comment by dragonmacher** [comment]
What you are suggesting seems like a hybrid case of when we have now.   The parallel decompiling we use now does not need caching, since our analyzers only ever examine a function one time.   The UI uses a cache since users tend to navigate back and forth while examining a program.

I'm assuming that your use case(s) involve revisiting functions that have already been decompiled?

**Comment by astrelsky** [comment]
> What you are suggesting seems like a hybrid case of when we have now. The parallel decompiling we use now does not need caching, since our analyzers only ever examine a function one time. The UI uses a cache since users tend to navigate back and forth while examining a program.
> 
> I'm assuming that your use case(s) involve revisiting functions that have already been decompiled?

Yes. I'm analyzing functions based on references to a classes vtable. The same function could be revisited hundreds of times.
